### PR TITLE
[FLINK-21479][coordination] Provide read-only interface of TaskManagerTracker to ResourceAllocationStrategy

### PR DESCRIPTION
**Pull request by KarmaGYZ**
## What is the purpose of the change

This is a predecessor for optimizing performance of ResourceAllocationStrategy (FLINK-21174).

To optimize the performance, we will need to build and maintain index for registered/pending resources. As the strategy is designed to be stateless, we propose to build and maintain the index at TaskManagerTracker, providing only access methods to the strategy.

To decouple index accessing from the common FineGrainedSlotManager workflow, while preventing the strategy from directly modifying the states, we can introduce a read-only interface of TaskManagerTracker and pass it to the strategy. In this way, we can easily extend the read-only interface to provide more index-accessing methods in future.

## Verifying this change

This change is already covered by existing tests.

### CODE REVIEW

**Comment by KarmaGYZ** [comment]
cc @xintongsong

File: flink-runtime/src/main/java/org/apache/flink/runtime/resourcemanager/slotmanager/FineGrainedSlotManager.java [hunk 1]
```diff
@@ -464,9 +464,7 @@ private void checkResourceRequirements() {
                                                                 .getDefaultSlotResourceProfile())));
```
Comment (xintongsong) [inline flink-runtime/src/main/java/org/apache/flink/runtime/resourcemanager/slotmanager/FineGrainedSlotManager.java:464]: `availableResources` can be removed

File: flink-runtime/src/test/java/org/apache/flink/runtime/resourcemanager/slotmanager/DefaultResourceAllocationStrategyTest.java [hunk 2]
```diff
@@ -49,56 +53,61 @@
 
     @Test
     public void testFulfillRequirementWithRegisteredResources() {
-        final Map<InstanceID, Tuple2<ResourceProfile, ResourceProfile>> registeredResources =
-                new HashMap<>();
+        final TaskManagerInfo taskManager =
+                new FineGrainedTaskManagerRegistration(
```
Comment (xintongsong) [inline flink-runtime/src/test/java/org/apache/flink/runtime/resourcemanager/slotmanager/DefaultResourceAllocationStrategyTest.java:57]: Maybe introduce a `TestingTaskManagerInfo`.

**Review by xintongsong — COMMENTED** [review]
Thanks @KarmaGYZ. PR looks quite good. I have only two minor comment.

**Comment by KarmaGYZ** [comment]
Thanks for the review @xintongsong . PR updated.

**Review by xintongsong — APPROVED** [review]
LGTM. Merging this once AZP gives green light.

**Pull request closed without merging**